src/plugins/workspace/server/saved_objects/workspace_saved_objects_client_wrapper.ts

```diff
@@ -616,23 +661,81 @@ export class WorkspaceSavedObjectsClientWrapper {
       return wrapperOptions.client;
     }
 
+    const ACLAuditDecorator = function <T extends (...args: any[]) => any>(
+      fn: T,
+      request: OpenSearchDashboardsRequest
+    ): T {
+      return function (...args: Parameters<T>): ReturnType<T> {
+        clientCallAuditor?.increment(CLIENT_CALL_AUDITOR_KEY.incoming);
+        const result = fn.apply(wrapperOptions.client, args);
+        if (result instanceof Promise) {
+          result.then(
+            () => {
+              clientCallAuditor?.increment(CLIENT_CALL_AUDITOR_KEY.outgoing);
+              try {
+                const checkoutInfo = JSON.stringify([fn.name, ...args]);
+                if (clientCallAuditor?.isAsyncClientCallsBalanced()) {
+                  ACLAuditor?.checkout(checkoutInfo);
+                }
+              } catch (e) {
+                if (clientCallAuditor?.isAsyncClientCallsBalanced()) {
+                  ACLAuditor?.checkout();
+                }
+              }
+            },
+            /**
+             * The catch here is required because unhandled promise will make server crashed,
+             * and we will reset the auditor state when catch an error.
+             */
+            () => {
```

Comment:
Do we need to return `Promise.reject` here? Or the promise will be resolved after return.

Reply:
No, we already return the result, the `result` will give the rejected error if the promise fails.